kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/service/RuleCapabilitiesServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+        }
+
+        org.kie.server.controller.api.model.spec.ContainerSpec containerSpec = serverTemplate.getContainerSpec(containerSpecKey.getId());
+
+        ContainerConfig containerConfig = containerSpec.getConfigs().get(Capability.RULE);
+        if (containerConfig == null) {
+            containerConfig = new RuleConfig();
+            containerSpec.getConfigs().put(Capability.RULE, containerConfig);
+        }
+
+        ((RuleConfig) containerConfig).setPollInterval(interval);
+        ((RuleConfig) containerConfig).setScannerStatus(KieScannerStatus.STARTED);
+
+        templateStorage.update(serverTemplate);
+
+        List<Container> containers = kieServerInstanceManager.startScanner(serverTemplate, containerSpec, interval);
+
+        notificationService.notify(serverTemplate, containerSpec, containers);
+    }
+
+    @Override
+    public void stopScanner(final ContainerSpecKey containerSpecKey) {
+        ServerTemplate serverTemplate = templateStorage.load(containerSpecKey.getServerTemplateKey().getId());
+        if (serverTemplate == null) {
+            throw new RuntimeException("No server template found for id " + containerSpecKey.getServerTemplateKey().getId());
+        }
+
+        org.kie.server.controller.api.model.spec.ContainerSpec containerSpec = serverTemplate.getContainerSpec(containerSpecKey.getId());
+
+        ContainerConfig containerConfig = containerSpec.getConfigs().get(Capability.RULE);
+        if (containerConfig == null) {
+            containerConfig = new RuleConfig();
+            containerSpec.getConfigs().put(Capability.RULE, containerConfig);
+        }
+
+        ((RuleConfig) containerConfig).setPollInterval(null);
+        ((RuleConfig) containerConfig).setScannerStatus(KieScannerStatus.STOPPED);
+
+        templateStorage.update(serverTemplate);
+
+        List<Container> containers = kieServerInstanceManager.stopScanner(serverTemplate, containerSpec);
+
+        notificationService.notify(serverTemplate, containerSpec, containers);
+    }
+
+    @Override
+    public void upgradeContainer(final ContainerSpecKey containerSpecKey, ReleaseId releaseId) {
+        ServerTemplate serverTemplate = templateStorage.load(containerSpecKey.getServerTemplateKey().getId());
+        if (serverTemplate == null) {
+            throw new RuntimeException("No server template found for id " + containerSpecKey.getServerTemplateKey().getId());
+        }
+
+        org.kie.server.controller.api.model.spec.ContainerSpec containerSpec = serverTemplate.getContainerSpec(containerSpecKey.getId());
+        if (releaseId.getGroupId() == null) {
+            releaseId.setGroupId(containerSpec.getReleasedId().getGroupId());
+        }
+        if (releaseId.getArtifactId()== null) {
+            releaseId.setArtifactId(containerSpec.getReleasedId().getArtifactId());
+        }
+
+        ((ContainerSpec)containerSpec).setReleasedId(releaseId);
```

Comment:
it's actually not needed, the case as containerSpec is already ContainerSpec so it's again left over after removing interfaces for data objects